src/embedding/ad.ts

```diff
@@ -0,0 +1,584 @@
+// (C) 2020 GoodData Corporation
+
+import isEmpty = require('lodash/isEmpty');
+import {
+    CommandFailed,
+    IObjectMeta,
+    IGdcMessageEvent,
+    IGdcMessageEventDataWrapper,
+    ICommandFailedBody
+} from './common';
+
+type ANALYTICAL_DESIGNER = 'analyticalDesigner';
+const AD_PRODUCT_NAME = 'analyticalDesigner';
+
+// TODO: these are part of client (gooddata-js), move to model (typings)
+export interface IBaseExportConfig {
+    title?: string;
+    format?: 'xlsx' | 'csv' | 'raw';
+    mergeHeaders?: boolean;
+}
+
+// TODO: these are part of client (gooddata-js), move to model (typings)
+// NOTE: AFM is ommitted on purpose; it should be added by AD itself; create new type using Omit<>
+export interface IExportConfig extends IBaseExportConfig {
+    showFilters?: boolean;
+}
+
+export namespace EmbeddedAnalyticalDesigner {
+    export type IADMessageEvent<T, TBody> = IGdcMessageEvent<ANALYTICAL_DESIGNER, T, TBody>;
+
+    export enum GdcAdCommandType {
+        Save = 'adSave',
+        SaveAs = 'adSaveAs',
+        Export = 'adExport',
+        Undo = 'adUndo',
+        Redo = 'adRedo'
+    }
+
+    export enum GdcAdEventType {
+        NewInsightInitialized = 'adNewInsightInitialized',
+        InsightOpened = 'adInsightOpened',
+        InsightSaved = 'adInsightSaved',
+        UndoFinished = 'adUndoFinished',
+        RedoFinished = 'adRedoFinished',
+        ExportFinished = 'adExportFinished'
+    }
+
+    export enum GdcAdErrorType {
+        /**
+         * The posted command is not recognized by AD.
+         */
+        InvalidCommand = 'adError:invalidCommand',
+
+        /**
+         * Argument specified in the command body is invalid; it has failed the syntactical or semantic
+         * validation done by AD.
+         */
+        InvalidArgument = 'adError:invalidArgument',
+
+        /**
+         * Command was posted when AD is not in a state to process the command. For instance:
+         *
+         * - trying to do save/save-as on new, empty insight
+         * - trying to do save/save-as on insight that is in error
+         * - trying to do undo when there is no step-back available
+         * - trying to do redo when there is no step-forward available
+         */
+        InvalidState = 'adError:invalidState',
+
+        /**
+         * The Unexpected Happened.
+         */
+        RuntimeError = 'adError:runtime'
+    }
+
+    /**
+     * TODO:
+     * This event will be emitted if AD runs into errors while processing the posted command.
+     *
+     * @remarks see {@link GdcAdErrorType} for types of errors that may fly
+     */
+    export type AdCommandFailed = CommandFailed<ANALYTICAL_DESIGNER, ICommandFailedBody<GdcAdErrorType>>;
+
+    export type AdCommandFailedData = IGdcMessageEventDataWrapper<
+        ANALYTICAL_DESIGNER,
+        'appCommandFailed',
+        ICommandFailedBody<GdcAdErrorType>
+    >;
+
+    export function createAdCommandFailedData(
+        correlationId: string,
```

Comment:
`correlationId ` should be rename to `contextId`

Reply:
i still prefer `correlationId` name. This name represent the corporation with outer app

Reply:
As discusstion [here](https://docs.google.com/document/d/1NXfCDgADOnUFPzc05yx38K3KdzHZnAZIvFksOCn2u68/edit?disco=AAAAJRbfXXg), I think we already have agreement on `contextId`